api/pkg/store/mongo/store.go

```diff
@@ -25,37 +25,58 @@ func NewStore(db *mongo.Database) *Store {
 
 func (s *Store) ListDevices(ctx context.Context, perPage, page int, filters []models.Filter) ([]models.Device, int, error) {
 	skip := perPage * (page - 1)
+	var queryMatch []bson.M
 	var queryFilter []bson.M
 	for _, filter := range filters {
-		if filter.Type != "property" {
-			continue
-		}
+		if filter.Type == "property" {
+			switch filter.PropertyParams.Operator {
+			case "like":
+				queryFilter = append(queryFilter, bson.M{
+					filter.PropertyParams.Name: bson.M{
+						"$regex":   filter.PropertyParams.Value,
+						"$options": "i",
+					},
+				})
+			case "eq":
+				queryFilter = append(queryFilter, bson.M{
+					filter.PropertyParams.Name: bson.M{
+						"$eq": filter.PropertyParams.Value,
+					},
+				})
+			case "bool":
+				operator, err := strconv.ParseBool(filter.PropertyParams.Value)
+				if err != nil {
+					return nil, 0, err
+				}
+				queryFilter = append(queryFilter, bson.M{
+					filter.PropertyParams.Name: bson.M{
+						"$eq": operator,
+					},
+				})
 
-		switch filter.Params.Operator {
-		case "like":
-			queryFilter = append(queryFilter, bson.M{
-				filter.Params.Name: bson.M{
-					"$regex":   filter.Params.Value,
-					"$options": "i",
-				},
-			})
-		case "eq":
-			queryFilter = append(queryFilter, bson.M{
-				filter.Params.Name: bson.M{
-					"$eq": filter.Params.Value,
-				},
-			})
-		case "bool":
-			operator, err := strconv.ParseBool(filter.Params.Value)
-			if err != nil {
-				return nil, 0, err
 			}
-			queryFilter = append(queryFilter, bson.M{
-				filter.Params.Name: bson.M{
-					"$eq": operator,
-				},
-			})
 		}
+		if filter.Type == "operator" {
```

Comment:
We have the switch case, this should be reworked and keep using it as it makes the code easier to read.

Please see if you can get this to work.

Reply:
It seems a cleaned code, preserving your logic, would be:

```diff
diff --git a/api/pkg/store/mongo/store.go b/api/pkg/store/mongo/store.go
index c320c7c..beddd6d 100644
--- a/api/pkg/store/mongo/store.go
+++ b/api/pkg/store/mongo/store.go
@@ -28,7 +28,8 @@ func (s *Store) ListDevices(ctx context.Context, perPage, page int, filters []mo
 	var queryMatch []bson.M
 	var queryFilter []bson.M
 	for _, filter := range filters {
-		if filter.Type == "property" {
+		switch filter.Type {
+		case "property":
 			switch filter.PropertyParams.Operator {
 			case "like":
 				queryFilter = append(queryFilter, bson.M{
@@ -53,26 +54,24 @@ func (s *Store) ListDevices(ctx context.Context, perPage, page int, filters []mo
 						"$eq": operator,
 					},
 				})
-
 			}
-		}
-		if filter.Type == "operator" {
+		case "operator":
+			operator := nil
 			switch filter.OperatorParams.Name {
 			case "and":
-
-				queryMatch = append(queryMatch, bson.M{
-					"$match": bson.M{"$and": queryFilter},
-				})
-				queryFilter = []bson.M{}
+				operator = "$and"
 			case "or":
-				queryMatch = append(queryMatch, bson.M{
-					"$match": bson.M{"$or": queryFilter},
-				})
-				queryFilter = []bson.M{}
+				operator = "$or"
 			}
-		}
 
+			queryMatch = append(queryMatch, bson.M{
+				"$match": bson.M{operator: queryFilter},
+			})
+
+			queryFilter = []bson.M{}
+		}
 	}
+
 	if len(queryFilter) > 0 {
 		queryMatch = append(queryMatch, bson.M{
 			"$match": bson.M{"$or": queryFilter},
```

However, it is not clear to me why you are resetting `queryFilter` when selecting the operators. It seems odd to me.

Reply:
QueryFilter is an intermediate slice to to append each filter until I got the operator
in this moment I append all values acumulated in QueryFilter to respective and/or groups in QueryMatch slice. 

To restart this routine I reset QueryFilter values.  @gustavosbarreto do you know another aproach to implement this task? This was my first